[NPoS] Unlimited number of nominators eligible for Rewards payout

Ank4n opened the pull request:
helps https://github.com/paritytech/polkadot-sdk/issues/439. 
closes https://github.com/paritytech/polkadot-sdk/issues/473.

polkadot companion: https://github.com/paritytech/polkadot/pull/6765.

Original PR: https://github.com/paritytech/substrate/pull/13059.

### Context
Rewards payout is processed today in a single block and limited to `MaxNominatorRewardedPerValidator`. This number is currently `512` on both Kusama and Polkadot. 

This PR tries to scale the nominators payout to an unlimited count in a multi-block fashion. Exposures are stored in pages, with each page capped to a certain number (`MaxExposurePageSize`). Starting out, this number would be the same as `MaxNominatorRewardedPerValidator`, but eventually, this number can be lowered through new runtime upgrades to limit the rewardeable nominators per dispatched call instruction.

The changes in the PR are backward compatible.

### Note about the release
~This PR introduces a new storage item `MaxExposurePageCount` which defaults to minimum value of `1` when not set. This means functionally, the pallet will continue to work exactly as before after this release. To enable paged rewards payout, `MaxExposurePageCount` can be increased to a bigger number by OpenGov for each runtime separately.~
Edit: `MaxExposurePageCount` limit is removed. The nominators are paged if they exceed the `MaxExposurePageSize` limit per validator.

### How payouts would look like after this change
Staking exposes two calls, 1) the existing `payout_stakers` and 2) `payout_stakers_by_page`.
#### payout_stakers
This remains backward compatible with no signature change. If for a given era a validator has multiple pages, they can call `payout_stakers` multiple times. The pages are executed in an ascending sequence and the runtime takes care of preventing double claims.
#### payout_stakers_by_page
Very similar to `payout_stakers` but also accepts an extra param `page_index`. An account can choose to payout rewards only for an explicitly passed `page`.

Lets look at an example scenario.
Given an active validator on Kusama had 1100 nominators, `MaxExposurePageCount` set to 10 and `MaxExposurePageSize` set to 512 for Era `e`. In order to pay out rewards to all nominators, the caller would need to call `payout_stakers` 3 times.
- `payout_stakers(origin, stash, e)` => will pay the first 512 nominators.
- `payout_stakers(origin, stash, e)` => will pay the second set of 512 nominators.
- `payout_stakers(origin, stash, e)` => will pay the last set of 76 nominators.
...
- `payout_stakers(origin, stash, e`) => calling it the 4th time would return an error `InvalidPage`.

The above calls can also be replaced by `payout_stakers_by_page` and passing a page explicitly.

#### Commission note
Validator commission is paid out in chunks across all the pages where each commission chunk is proportional to the total stake of the current page. This implies higher the total stake of a page, higher will be the commission. If all the pages of a validator's single era are paid out, the commission paid to the validator will be the same as if it were a non-paged exposure.

### Migration Note
Strictly speaking, we did not need to bump our storage version since there is no migration of storage in this PR. But it is still useful to mark a storage upgrade for the following reasons
- New storage items are introduced while some older storage items are deprecated.
- For the next `HistoryDepth` eras, the exposure would be incrementally migrated to its corresponding paged storage item.
- Runtimes using staking pallet would strictly need to wait at least `HistoryDepth` eras with current upgraded version (14). At some era `E` such that `E > era_at_which_V14_gets_into_effect + HistoryDepth`, we will upgrade to version `X` which will clean up/kill the deprecated storage items. 
In other words, it is a strict requirement that E<sub>x</sub> - E<sub>14</sub> > `HistoryDepth`, where
E<sub>x</sub> = Era at which deprecated storages are removed from runtime,
E<sub>14</sub> = Era at which runtime is upgraded to version 14.
- For Polkadot and Kusama, there is a [tracker ticket ](https://github.com/paritytech/polkadot-sdk/issues/433) to clean up the deprecated storage items.

### Storage Changes
#### Added
- `ErasStakersOverview`
- `ClaimedRewards`
- `ErasStakersPaged`

#### Deprecated
The following can be cleaned up after 84 eras. [Tracker issue for cleaning up](https://github.com/paritytech/polkadot-sdk/issues/433).
- `ErasStakers`.
- `ErasStakersClipped`.
- `StakingLedger.claimed_rewards`, renamed to `StakingLedger.legacy_claimed_rewards`.

### Config Changes
- Renamed `MaxNominatorRewardedPerValidator` to `MaxExposurePageSize`.
- ~New item `MaxExposurePageCount`~


### TODO 
- [x] Tracker ticket for cleaning up the old code after 84 eras.
- [x] Add companion.
- [ ] Redo benchmarks before merge.
- [x] Add Changelog for pallet_staking.
- [x] Pallet should be configurable to enable/disable paged rewards.
- [x] Commission payouts are distributed across pages.
- [x] Review documentation thoroughly.
- [x] Rename `MaxNominatorRewardedPerValidator` -> `MaxExposurePageSize`.
- [x] NMap for ErasStakersPaged.
- [x] Deprecate ErasStakers.
- [x] Integrity tests.

## Followup
- [Runtime api for deprecated ErasStakers storage item.](https://github.com/paritytech/polkadot-sdk/issues/426)

Ank4n committed:
rewards payout in multi block fashion

Ank4n committed:
rewards payout in multi block fashion

Ank4n committed:
use MaxExposurePageSize from staking

Ank4n committed:
use MaxExposurePageSize from staking

Ank4n committed:
add max exposure page count to staking interface

Ank4n committed:
add max exposure page count to staking interface

Ank4n committed:
fmt

Ank4n committed:
fmt

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2464837

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2464838

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2464836

Ank4n committed:
fix unimplemented function

Ank4n committed:
fix unimplemented function

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2464935

Ank4n committed:
feature gate

Ank4n committed:
feature gate

Ank4n committed:
add staking api

Ank4n committed:
Merge branch 'master' into ankan/paged-rewards-rebased2

Ank4n commented:
bot rebase

Ank4n committed:
Merge branch 'ankan/paged-rewards-staking-api' into ankan/paged-rewards-rebased2

Ank4n committed:
fmt

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2514108

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2514101

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2514102

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2514103

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2514257

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2514250

Ank4n committed:
update weights

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2514375

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2514376

Ank4n committed:
fix imports

Ank4n committed:
fix docs

Ank4n committed:
Merge branch 'master' into ankan/paged-rewards-rebased2

Ank4n commented:
bot rebase

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2548419

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2548420

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2548421

Ank4n committed:
fix merge

kianenigma commented on bin/node/runtime/src/lib.rs:
```diff
@@ -582,7 +582,8 @@ impl pallet_staking::Config for Runtime {
 	type EraPayout = pallet_staking::ConvertCurve<RewardCurve>;
 	type NextNewSession = Session;
-	type MaxNominatorRewardedPerValidator = MaxNominatorRewardedPerValidator;
+	type MaxExposurePageSize = MaxExposurePageSize;
```
```suggestion
	type MaxExposurePageSize = ConstU32<256>;
```

kianenigma commented on bin/node/runtime/src/lib.rs:
```diff
@@ -605,7 +606,7 @@ impl pallet_fast_unstake::Config for Runtime {
 	type MaxErasToCheckPerBlock = ConstU32<1>;
 	#[cfg(feature = "runtime-benchmarks")]
-	type MaxBackersPerValidator = MaxNominatorRewardedPerValidator;
+	type MaxExposurePageSize = <Runtime as pallet_staking::Config>::MaxExposurePageSize;
```
don't have a strong opinion now, but this could be moved to `trait StakingInterface`, as a feature gated type item.

Ank4n replied:
Good catch actually. There is already a `fn max_exposure_page_size()` in StakingInterface and this config item was unused.

kianenigma commented on bin/node/runtime/src/lib.rs:
```diff
@@ -2049,10 +2050,18 @@ impl_runtime_apis! {
 			Staking::api_nominations_quota(balance)
 		}
+
+		fn eras_stakers(era: sp_staking::EraIndex, account: AccountId) -> pallet_staking::Exposure<AccountId, Balance> {
+			Staking::api_eras_stakers(era, account)
+		}
```
now's your last time to ask yourself: is the (legacy) name `eras_stakers` actually good? shoudl we not use the oppurtunity and rename it to something better? :D

Ank4n replied:
I don't have a strong opinion here. 

On one hand, since this is meant for backward compatibility and all similar storage items are named similarly (eras_stakers, eras_stakers_paged, eras_stakers_overview ...), it makes sense to keep a similar name. 

We could call it `eras_exposure` as well but then we are exposing a new term (exposure) outside staking world which may be just an internal concept. `Eras Stakers` might not be such a bad public name (staker info for the era).

Do you have a name suggestion?

kianenigma commented on frame/fast-unstake/src/lib.rs:
```diff
@@ -142,7 +142,7 @@ pub mod pallet {
 		/// Use only for benchmarking.
 		#[cfg(feature = "runtime-benchmarks")]
-		type MaxBackersPerValidator: Get<u32>;
+		type MaxExposurePageSize: Get<u32>;
```
Still think this is part of `StakingInterface`.

Ank4n replied:
Removed 👍

kianenigma commented on frame/nomination-pools/src/mock.rs:
```diff
@@ -55,16 +55,30 @@ impl sp_staking::StakingInterface for StakingMock {
 	}
 
-	fn desired_validator_count() -> u32 {
+	fn stash_by_ctrl(_controller: &Self::AccountId) -> Result<Self::AccountId, DispatchError> {
```
nit: diff could be made better by avoiding re-orderings.

Ank4n replied:
I fixed the reordering to realise why I did that in the first place.
My IDE nags that there is a`Different impl member order from the trait `. :D

kianenigma commented on frame/staking/runtime-api/Cargo.toml:
```diff
@@ -15,10 +15,14 @@ targets = ["x86_64-unknown-linux-gnu"]
 codec = { package = "parity-scale-codec", version = "3.0.0", default-features = false, features = ["derive"] }
 sp-api = { version = "4.0.0-dev", default-features = false, path = "../../../primitives/api" }
+sp-staking = { version = "4.0.0-dev", default-features = false, path = "../../../primitives/staking" }
+pallet-staking = { version = "4.0.0-dev", default-features = false, path = "../../staking" }
```
It is not a strict need, but technically we should move `Exposure`, or a type similar to it into `sp_staking` and not import the whole pallet staking here.

This API is supposed to be unbiased.

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -674,6 +693,13 @@ impl<T: Config> Pallet<T> {
 		<ErasStakersClipped<T>>::remove_prefix(era_index, None);
 		#[allow(deprecated)]
 		<ErasValidatorPrefs<T>>::remove_prefix(era_index, None);
+		#[allow(deprecated)]
```
Do we properly benchmark this `clear_era_information` and report its weight anywhere?

Ank4n replied:
Not really. We don't benchmark anything around triggering of new era but return max weight.

https://github.com/paritytech/substrate/blob/master/frame/session/src/lib.rs#L571

kianenigma replied:
😢

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -732,6 +741,140 @@ impl<AccountId, Balance: Default + HasCompact> Default for Exposure<AccountId, B
+
+		for chunk in individual_chunks {
+			let mut page_total: Balance = Zero::zero();
+			let mut others: Vec<IndividualExposure<AccountId, Balance>> = vec![];
```
```suggestion
			let mut others: Vec<IndividualExposure<AccountId, Balance>> = Vec::with_capacity(chunks.len());
```

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -732,6 +741,140 @@ impl<AccountId, Balance: Default + HasCompact> Default for Exposure<AccountId, B
+			let mut page_total: Balance = Zero::zero();
+			let mut others: Vec<IndividualExposure<AccountId, Balance>> = vec![];
+			for individual in chunk.iter() {
+				page_total = page_total.saturating_add(individual.value);
```
```suggestion
				page_total.saturating_accrue(individual.value);
```

kianenigma commented on frame/staking/src/migrations.rs:
```diff
@@ -52,6 +52,47 @@ impl Default for ObsoleteReleases {
 #[storage_alias]
 type StorageVersion<T: Config> = StorageValue<Pallet<T>, ObsoleteReleases, ValueQuery>;
 
+pub mod v14 {
```
explanation?

kianenigma commented on frame/staking/src/migrations.rs:
```diff
@@ -52,6 +52,47 @@ impl Default for ObsoleteReleases {
+
+	pub struct MigrateToV14<T>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV14<T> {
+		fn on_runtime_upgrade() -> Weight {
```
Please https://github.com/paritytech/substrate/issues/13107#issuecomment-1443895774? :D

Ank4n commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into ankan/paged-rewards-rebased2

Ank4n commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into ankan/paged-rewards-rebased2

Ank4n committed:
resolve pr comments

Ank4n committed:
keep the order of staking interface impl

Ank4n committed:
move Exposure, ExposurePage, ExposureOverview and IndividualExposure to sp-staking

Ank4n committed:
remove dependency to stakign for staking runtime api

Ank4n committed:
initialise size of others while prepping pages

Ank4n committed:
imrpovements

Ank4n committed:
fix code

Ank4n committed:
link migration changelog

Ank4n committed:
remove unused import

Ank4n committed:
fix staking tests

Ank4n committed:
fmt

gpestana reviewed (dismissed):
Left a few nit comments, overall it's looking good. I have a design question though. You wrote:

> -   Validator commission is paid out in chunks across all the pages where each commission chunk is proportional to the total stake of the current page. This encourages the validator to payout all the exposure pages.

I would argue that if the commission is proportional total stake of the current page, this actually disincentives the validator to call `payout_stakers` to all pages, especially if the tail page has very few rewards. How about the commission paid per page is proportional to the number of pages? i.e. `total_rewards/n_pages` . This would incentivise the validator to payout the last page too, even if it has a small number of nominators exposed.

Ank4n committed:
Apply suggestions from code review

Co-authored-by: Gonçalo Pestana <g6pestana@gmail.com>

Commit:
Merge remote-tracking branch 'origin/master' into ankan/paged-rewards-rebased2

Ank4n committed:
move EraInfo to lib

Ank4n commented on frame/staking/src/pallet/mod.rs:
```diff
@@ -558,6 +641,217 @@ pub mod pallet {
 	#[pallet::getter(fn current_planned_session)]
 	pub type CurrentPlannedSession<T> = StorageValue<_, SessionIndex, ValueQuery>;
 
+	/// Wrapper struct for Era related information. It is not a pure encapsulation as these storage
```
Moved

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -112,11 +112,17 @@
+//! number of nominators exceeds [`Config::MaxExposurePageSize`], then the nominators are stored in
+//! multiple pages of [`Config::MaxExposurePageSize`], each with up to a maximum of
+//! [`Config::MaxExposurePageCount`] pages. To pay out all nominators, `payout_stakers` must be
+//! called once for each available page. In a scenario where the number of nominators N exceed M, where M
```
Don't ask me how I know, but this and the next line exceed 100 chars :D

Ank4n replied:
🙈

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -939,6 +961,217 @@ where
+			return None
+		}
+
+		let overview = overview.unwrap();
```
```suggestion
		let overview = overview.expect("checked above; qed");
```

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -939,6 +961,217 @@ where
+
+	/// Get exposure info for a validator at a given era and page.
+	///
+	/// This builds a paged exposure from `ExposureOverview` and `ExposurePage` of the
```
Suggesting something like https://marketplace.visualstudio.com/items?itemName=stkb.rewrap

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -939,6 +961,217 @@ where
+
+	/// Get exposure info for a validator at a given era and page.
+	///
+	/// This builds a paged exposure from `ExposureOverview` and `ExposurePage` of the
```
suggesting https://marketplace.visualstudio.com/items?itemName=stkb.rewrap

Ank4n replied:
IntelliJ user :D ..

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -939,6 +961,217 @@ where
+			return ErasStakers::<T>::get(era, validator)
+		}
+
+		let overview = overview.unwrap();
```
expect

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -939,6 +961,217 @@ where
+
+		let overview = overview.unwrap();
+
+		if overview.page_count == 0 {
```
would have been fine without it as well, right? can stay, just asking.

Ank4n replied:
Yes good catch. Removed it.

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -939,6 +961,217 @@ where
+		<ErasStakersOverview<T>>::get(&era, validator)
+			.map(|overview| {
+				if overview.page_count == 0 && overview.own > Zero::zero() {
+					// this means the validator has their own stake but no nominators
```
quite confusing why if a validator only has their self-stake, this function returns 1. Why we ended up coding it like this? I suspect the call site might have unreasonable expectations that leaked here.

Ank4n replied:
Good point. I agree its not clean and will fix it.

The `call` to payout stakers pays the validator stake only when the page is 1. But as you pointed out it should be handled by the calling function to take care of paying out the validator stake in all these cases instead of an underlying function lying about the page count.

Ank4n replied:
On further digging, may be this is not too bad. With this PR, the expectation is that the caller has to specify the page to claim rewards for. If a validator has no nominators (therefore no nominator pages), but their own stake is exposed, that still needs to be paid in a page, hence one page of exposure.

I think the fact that we differentiate between own and other stake of the validator adds to this confusion. I have added a little bit more explanation in the comment for this.

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -939,6 +961,217 @@ where
+			exposure
+		};
+
+		let (exposure_overview, exposure_pages) = exposure.into_pages(page_size);
```
```suggestion
		let (exposure_overview, exposure_pages) = exposure.into_pages(page_size);
		debug_assert_eq!(exposure_pages.len(), required_page_count);
```

or something like `defensive_ensure!()`?

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -939,6 +961,217 @@ where
+		let (exposure_overview, exposure_pages) = exposure.into_pages(page_size);
+
+		<ErasStakersOverview<T>>::insert(era, &validator, &exposure_overview);
+		exposure_pages.iter().enumerate().for_each(|(page, paged_exposure)| {
```
```suggestion
		exposure_pages.iter().take(max_page_count).enumerate().for_each(|(page, paged_exposure)| {
```

Both of my suggestions here are super defensive, maybe you should double check `exposure_pages` instead.

Ank4n replied:
I added a defensive assert above on `exposure_pages`. This step does not need to be defensive since nothing will panic here.

kianenigma commented on frame/staking/src/pallet/mod.rs:
```diff
@@ -1464,21 +1541,42 @@ pub mod pallet {
+		/// and vice versa. If rewards are not claimed in [`Config::HistoryDepth`] eras, they are
+		/// lost.
+		///
+		/// # <weight>
```
No need for the #weight stuff, they are being deprecated.

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -134,21 +130,45 @@ impl<T: Config> Pallet<T> {
+					.with_weight(T::WeightInfo::payout_stakers_alive_staked(0))
+			})?;
+
+		Self::do_payout_stakers_by_page(validator_stash, era, page)
```
You can pass in `ledger` here as well and reuse it.

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -159,29 +179,28 @@ impl<T: Config> Pallet<T> {
-			.claimed_rewards
+			.legacy_claimed_rewards
 			.retain(|&x| x >= current_era.saturating_sub(history_depth));
+		<Ledger<T>>::insert(&controller, &ledger);
```
In the `TargetList` line of work, we are creating a VERY IMPORTANT assumption that any time a `ledger`'s stake fields change, we call into `update_ledger`.

We should first fix this by making the `Ledger` have a private fileds, and only specific `update` methods that hold these assumptions.

This code is fone though, because we are only touching `legacy_claimed_rewards`.

If you follow the code, you will see that in `make_payout`, we call `make_payout`, which is right, but is a thin thread, if you know what I mean.

Ank4n replied:
Also somewhat supports the idea that claimed_rewards does not belong to the ledger.

kianenigma reviewed (approved):
Everything looks good, just haven't read the tests in detail yet. 

Audit can certainly begin. 

I suggest setting up some accounts such that as soon as this PR is in a release and enacted in westend, we test it live. Same for Kusama.

Ank4n commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into ankan/paged-rewards-rebased2

Ank4n committed:
fix docs

Ank4n committed:
expect instead of unwrap

Ank4n committed:
remove useless code

Ank4n committed:
remove deprecated weight doc

Ank4n committed:
clip required page count

Ank4n committed:
defensively assert

Ank4n committed:
doc wrapping fix

Ank4n committed:
explain why we return one page when no nominator pages exist

Ank4n commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into ankan/paged-rewards-rebased2

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2688571

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2688556

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2688557

Ank4n committed:
eras stakers might be an unsafe runtime api, should evaluate separately later

Ank4n committed:
remove eras stakers runtime api from node runtime

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2688625

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2688611

Ank4n committed:
rename Era info functions

Ank4n committed:
fmt

Ank4n committed:
Merge branch 'master' into ankan/paged-rewards-rebased2

Ank4n committed:
fix tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2857636

Ank4n committed:
fixes

Ank4n commented:
bot fmt

Commit:
".git/.scripts/commands/fmt/fmt.sh"

Ank4n commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into ankan/paged-rewards-rebased2

Ank4n committed:
Merge branch 'master' into ankan/paged-rewards-rebased2

Ank4n commented:
bot rebase

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2919488

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2919489

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2919487

Ank4n committed:
add back missed config while resolving conflict

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2919799

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2919800

Ank4n committed:
not needed anymore

Ank4n committed:
fix migration

rossbulat commented on bin/node/runtime/src/lib.rs:
```diff
@@ -2183,10 +2181,14 @@ impl_runtime_apis! {
 			Staking::api_nominations_quota(balance)
 		}
+
+		fn era_page_count(era: sp_staking::EraIndex, account: AccountId) -> sp_staking::PageIndex {
+			Staking::api_era_page_count(era, account)
+		}
```
This API call is only for validator exposures so the naming could reflect this and be a bit more intuitive. `era_page_count` gives me the impression that this is returning the total pages for a whole era.

```suggestion
		fn validator_exposure_page_count(era: sp_staking::EraIndex, who: AccountId) -> sp_staking::PageIndex {
			Staking::api_era_page_count(era, who)
		}
```

Ank4n replied:
The storage item that page corresponds to is called `ErasStakersPaged` (which derives its name from the existing non-paged `ErasStakers`). What do you think about renaming it as `eras_stakers_page_count`?

Ank4n replied:
Renamed to eras_stakers_page_count to keep in line with how storage items are named.

rossbulat commented on frame/fast-unstake/src/benchmarking.rs:
```diff
@@ -74,7 +74,7 @@ fn setup_staking<T: Config>(v: u32, until: EraIndex) {
 
 	for era in 0..=until {
-		let others = (0..T::MaxBackersPerValidator::get())
+		let others = (0..T::Staking::max_exposure_page_size())
```
Is this benchmark only considering a page size of 1? If we go multi page, with `MaxExposurePageCount` > 1, then would we need a double iterator for iterating the actual `page` count and `others` within them?

rossbulat replied:
Clarified that a call only deals with 1 page so multi-page not needed.

rossbulat commented on frame/staking/README.md:
```diff
@@ -92,11 +92,11 @@ An account can become a nominator via the [`nominate`](https://docs.rs/pallet-st
 valid behavior_ while _punishing any misbehavior or lack of availability_.
 
-Rewards must be claimed for each era before it gets too old by `$HISTORY_DEPTH` using the
+Rewards must be claimed for each era before it gets too old by [`HistoryDepth`] using the
 `payout_stakers` call. Any account can call `payout_stakers`, which pays the reward to the
```
```suggestion
Rewards must be claimed on a per-era basis via the `payout_stakers` call, and must be claimed before [`HistoryDepth`] eras have passed since the reward's inclusion, after which any unclaimed rewards will expire and be removed from storage. Any account can call `payout_stakers`, which pays the reward to the
```

rossbulat commented on frame/staking/src/mock.rs:
```diff
@@ -804,3 +814,7 @@ pub(crate) fn staking_events_since_last_call() -> Vec<crate::Event<Test>> {
 	(Balances::free_balance(who), Balances::reserved_balance(who))
 }
+
+pub(crate) fn allow_paged_rewards(max_pages: u32) {
```
```suggestion
pub(crate) fn set_max_exposure_page_count(max_pages: u32) {
```

rossbulat commented on frame/staking/src/pallet/mod.rs:
```diff
@@ -1455,21 +1531,30 @@ pub mod pallet {
+		/// them, then the list of nominators is paged, with each page being capped at
+		/// [`Config::MaxExposurePageSize`]. If a validator has more than one page of
+		/// nominators, the call needs to be made for each page separately in order for all the
+		/// nominators backing a validator receive the reward. The nominators are not sorted across
```
```suggestion
		/// nominators backing a validator to receive the reward. The nominators are not sorted across
```

rossbulat commented on frame/staking/src/pallet/mod.rs:
```diff
@@ -1771,6 +1856,35 @@ pub mod pallet {
+		/// them, then the list of nominators is paged, with each page being capped at
+		/// [`Config::MaxExposurePageSize`.] If a validator has more than one page of nominators,
+		/// the call needs to be made for each page separately in order for all the nominators
+		/// backing a validator receive the reward. The nominators are not sorted across pages and
```
```suggestion
		/// backing a validator to receive the reward. The nominators are not sorted across pages and
```

rossbulat commented on frame/staking/src/pallet/mod.rs:
```diff
@@ -1455,21 +1531,30 @@ pub mod pallet {
+		///
+		/// If a validator has more than [`Config::MaxExposurePageSize`] nominators backing
+		/// them, then the list of nominators is paged, with each page being capped at
+		/// [`Config::MaxExposurePageSize`]. If a validator has more than one page of
+		/// nominators, the call needs to be made for each page separately in order for all the
+		/// nominators backing a validator receive the reward. The nominators are not sorted across
+		/// pages and so it should not be assumed the highest staker would be on the topmost page
```
This is repeated at `payout_stakers_by_page` - perhaps remove here.
```suggestion
```

rossbulat commented on frame/babe/src/mock.rs:
```diff
@@ -195,7 +195,8 @@ impl pallet_staking::Config for Test {
 	type UnixTime = pallet_timestamp::Pallet<Test>;
 	type EraPayout = pallet_staking::ConvertCurve<RewardCurve>;
-	type MaxNominatorRewardedPerValidator = ConstU32<64>;
+	type MaxExposurePageSize = ConstU32<64>;
+	type MaxExposurePageCount = ConstU32<1>;
```
Would there be any issues making these `ConstU16`? 32 bits may be overkill for these numbers.

Ank4n replied:
An update, I want to bound EraStakersPaged and ClaimedRewards with these two values and boundedVec needs to be bounded by a u32 value. I will undo the changes to make it u16.

rossbulat commented on primitives/staking/src/lib.rs:
```diff
@@ -215,4 +224,121 @@ pub trait StakingInterface {
 	fn set_current_era(era: EraIndex);
 }
 
+/// The amount of exposure (to slashing) than an individual nominator has.
```
```suggestion
/// The amount of exposure for an era than an individual nominator has (susceptible to slashing).
```

rossbulat commented on primitives/staking/src/lib.rs:
```diff
@@ -215,4 +224,121 @@ pub trait StakingInterface {
 
+/// The amount of exposure (to slashing) than an individual nominator has.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct IndividualExposure<AccountId, Balance: HasCompact> {
```
The `IndividualExposure` struct is used project-wide only with `others` within an `Exposure`, so shall we just call this struct `OtherExposure` instead? We can then unify a lot of naming, such as renaming things like `expected_individual_exposures` to just `expected_others`.

Ank4n replied:
I actually think the name makes sense. In the context of a validator, this exposure set is called `others` since its other than validator's own stake, but in itself, it represents individual exposure of each nominator.

rossbulat commented on primitives/staking/src/lib.rs:
```diff
@@ -215,4 +224,121 @@ pub trait StakingInterface {
+	Default,
+	MaxEncodedLen,
+)]
+pub struct ExposureOverview<Balance: HasCompact + codec::MaxEncodedLen> {
```
The name `ExposureHead` is coming to mind here to hold this exposure metadata. Thoughts?

gpestana replied:
Or even more explicitly `ExposureMetadata`

Ank4n replied:
Renamed to `PagedExposureMetadata` since it is only present for Paged Exposure.

rossbulat commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -140,21 +140,45 @@ impl<T: Config> Pallet<T> {
 		validator_stash: T::AccountId,
 		era: EraIndex,
+	) -> DispatchResultWithPostInfo {
+		let controller = Self::bonded(&validator_stash).ok_or_else(|| {
+			Error::<T>::NotStash.with_weight(T::WeightInfo::payout_stakers_alive_staked(0))
+		})?;
+		let ledger = <Ledger<T>>::get(&controller).ok_or(Error::<T>::NotController)?;
```
Will go away once controllers are deprecated, so just a a short-term weight bump.

rossbulat replied:
Will go away once controllers are deprecated, so just a a short-term weight bump.

```suggestion
		let controller = Self::bonded(&validator_stash).ok_or_else(|| {
			Error::<T>::NotStash.with_weight(T::WeightInfo::payout_stakers_alive_staked(0))
		})?;
		let ledger = <Ledger<T>>::get(&controller).ok_or(Error::<T>::NotController)?;
```

rossbulat commented on frame/staking/src/lib.rs:
```diff
@@ -692,32 +702,50 @@ pub struct Nominations<T: Config> {
+/// This is useful where we need to take into account the validator's own stake and total exposure
+/// in consideration, in addition to the individual nominators backing them.
+#[derive(Encode, Decode, RuntimeDebug, TypeInfo, PartialEq, Eq)]
+struct ExposureExt<AccountId, Balance: HasCompact + codec::MaxEncodedLen> {
```
Since `ExposureExt` is only used to return `get_paged_exposure`, would `PagedExposure` be a more intuitive name for this struct?

Ank4n replied:
Honestly I struggled to name this. Its more of a wrapper struct, so I could potentially rename it to `PagedExposure` as you suggested. Other names that come to my mind is `PagedExposureOuter` or `PagedExposureFacade`.

Ank4n replied:
Renamed to `PagedExposure`. Looks more intuitive now I think. Thanks.

```rust
struct PagedExposure<AccountId, Balance> {
	exposure_metadata: PagedExposureMetadata<Balance>,
	exposure_page: ExposurePage<AccountId, Balance>,
}
```

Ank4n commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into ankan/paged-rewards-rebased2

Ank4n committed:
Update frame/staking/runtime-api/src/lib.rs

Co-authored-by: Ross Bulat <ross@parity.io>

Ank4n committed:
Update frame/staking/src/pallet/mod.rs

Co-authored-by: Ross Bulat <ross@parity.io>

Ank4n commented:
> Left a few nit comments, overall it's looking good. I have a design question though. You wrote:
> 
> > * Validator commission is paid out in chunks across all the pages where each commission chunk is proportional to the total stake of the current page. This encourages the validator to payout all the exposure pages.
> 
> I would argue that if the commission is proportional total stake of the current page, this actually disincentives the validator to call `payout_stakers` to all pages, especially if the tail page has very few rewards. How about the commission paid per page is proportional to the number of pages? i.e. `total_rewards/n_pages` . This would incentivise the validator to payout the last page too, even if it has a small number of nominators exposed.

I have updated the PR description now (to what we talked privately about). We don't want to necessarily enforce pay out of all page, and commission should still be based on how important the payout of the page is (total stake of the page). Also since reward payouts are permissionless, anybody can claim rewards for a page by paying the fee.

Ank4n committed:
Merge branch 'master' into ankan/paged-rewards-rebased2

Ank4n commented:
bot rebase

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/revision-of-rewarddestination-to-account-for-split-and-controller-removal/3360/2

Ank4n committed:
Merge branch 'master' into ankan/paged-rewards-rebased2

Ank4n committed:
cargo fmt

Ank4n commented:
bot fmt

Ank4n committed:
rename page count api

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153475

Ank4n committed:
make exposure page count and page size u16

Ank4n committed:
Apply suggestions from code review

Co-authored-by: Ross Bulat <ross@parity.io>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153556

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153554

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153638

Ank4n committed:
wrap to 100 letters

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153640

Ank4n committed:
Update frame/staking/src/lib.rs

Co-authored-by: Gonçalo Pestana <g6pestana@gmail.com>

Ank4n committed:
rustdoc for Individual exposure

Ank4n committed:
rename ExposureOverview to more explicit PagedExposureMetadata

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153826

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153825

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153824

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153876

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153875

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153877

Ank4n committed:
don't repeat docs

Ank4n committed:
rename the exposure_overfiew field

Ank4n committed:
rename ExposureExt

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153943

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153945

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153944

Ank4n committed:
fix benchmark

Ank4n commented:
bot fmt

Commit:
".git/.scripts/commands/fmt/fmt.sh"

Ank4n committed:
Trigger Build

Ank4n committed:
rename test fn to set max page count

Ank4n committed:
revert changing types of exposure page size and count back to u32

Ank4n committed:
Merge branch 'master' into ankan/paged-rewards-rebased2

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3359543

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3359541

muharem committed:
Pallets: Treasury deprecate `propose_spend` dispatchable (#14538)

* treasury deprecate dispatchables

* allow deprecated

* allow deprecated for benchmarks

* allow deprecated in tests

* allow deprecated for bounties tests

* deprecation month

jsidorenko committed:
Sync NFT metadata limits with AssetHub values (#14748)

* Sync NFT metadata limits with AssetHub values

* Use exact values

Ank4n committed:
unused import

Ank4n commented:
bot fmt

Commit:
".git/.scripts/commands/fmt/fmt.sh"

acatangiu committed:
sc-cli: add no-beefy flag to cli config (#14754)

Signed-off-by: Adrian Catangiu <adrian@parity.io>

acatangiu committed:
sc-consensus-beefy: fix initialization when BEEFY genesis state unavailable (#14752)

When BEEFY voter is initialized from scratch (no aux db persistent data present),
it needs to find BEEFY genesis block and all subsequent Mandatory blocks and
sync justifications for them.

The initialization code was getting active validator sets for these older blocks
from state, but in cases such as 'fast sync', state is unavailable.

This commit adds a fallback initialization mechanism when state is unavailable:
parse header Digests looking for validator set change log deposits.

Signed-off-by: Adrian Catangiu <adrian@parity.io>

kianenigma committed:
make `BagsList::put_in_front_of` be permissionless (#14714)

* make BagsList::put_in_fron_of be permissionless

* Update frame/bags-list/src/lib.rs

Co-authored-by: Liam Aharon <liam.aharon@hotmail.com>

* improve docs as well

* update lock

* Update frame/bags-list/Cargo.toml

Co-authored-by: Sam Johnson <sam@durosoft.com>

* fix

* Update frame/bags-list/src/lib.rs

Co-authored-by: Michael Assaf <94772640+snowmead@users.noreply.github.com>

---------

Co-authored-by: Liam Aharon <liam.aharon@hotmail.com>
Co-authored-by: Sam Johnson <sam@durosoft.com>
Co-authored-by: Michael Assaf <94772640+snowmead@users.noreply.github.com>

pmikolajczyk41 committed:
[contracts] Derive useful traits for public types (#14723)

* Types in the pallet

* Primitives

kianenigma committed:
add `frame_system::DefaultConfig` to individual pallet `DefaultConfigs` (#14453)

* add frame_system::DefaultConfig to individual pallet DefaultConfigs

* Fixes tests

* Minor fix

* ".git/.scripts/commands/fmt/fmt.sh"

* Adds UI Tests

---------

Co-authored-by: Nikhil Gupta <17176722+gupnik@users.noreply.github.com>
Co-authored-by: command-bot <>

gupnik committed:
Fixes CI (#14763)

muse254 committed:
fix: node cli docs (#14718)

Commit:
[fix lint warnings: Uniques pallet] fix clippy::missing_errors_doc lint warnings (#14588)

* Fix clippy::missing_errors_doc lint warnings

* update error docs

* cargo +nightly fmt

* add intra doc linking

* fmt

---------

Co-authored-by: parity-processbot <>

gitofdeepanshu committed:
Add `try_state` check to Pallet `MessageQueue` (#13502)

* feat: Add try_state for message_queue

* Update frame/message-queue/src/lib.rs

change if let to while let and modify bump_service_head function

Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* feat: update try_state, add checks for storage

* fix: add try_state builder for remaining tests

* Update frame/message-queue/src/mock.rs

Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* chore: assert statement to ensure

* Fix tests

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Use ensure instead of assert

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Fix function signature and feature gate

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Cleanup code

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

---------

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>
Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>
Co-authored-by: Juan <juangirini@gmail.com>

naterarmstrong committed:
Remove missing sentence (#14769)

Commit:
[fix lint warnings: NFTs pallet] fix clippy::missing_errors_doc lint warnings (#14648)

* fix missing errors doc warnings

* cargo +nightly fmt

* Update frame/nfts/src/features/create_delete_item.rs

* Update frame/nfts/src/features/create_delete_item.rs

* Update frame/nfts/src/features/transfer.rs

* Update frame/nfts/src/features/create_delete_collection.rs

* add intra doc linking for errors

* fmt

* Apply suggestions from code review

Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

---------

Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

lexnv committed:
chainHead: Limit ongoing operations (#14699)

* chainHead/api: Make storage/body/call pure RPC methods

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead: Add mpsc channel between RPC methods

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/subscriptions: Extract mpsc::Sender via BlockGuard

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/subscriptions: Generate and provide the method operation ID

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead: Generate `chainHead_body` response

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead: Generate `chainHead_call` response

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead: Generate `chainHead_storage` responses

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead: Propagate responses of methods to chainHead_follow

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/tests: Adjust `chainHead_body` responses

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/tests: Adjust `chainHead_call` responses

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/tests: Adjust `chainHead_call` responses

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/tests: Ensure unique operation IDs across methods

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/events: Remove old method events

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/subscriptions: Add limit helper

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/subscription: Expose limits to `BlockGuard`

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/tests: Adjust testing to ongoing operations

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead: Make limits configurable via `ChainHeadConfig`

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/tests: Adjust testing to `ChainHeadConfig`

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/tests: Ensure operation limits discards items

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead: Improve documentation

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead: Rename `OngoingOperations` -> `LimitOperations`

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead: Rename reserve -> reserve_at_most

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead: Use duration const instead of u64

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chainHead/subscription: Use tokio::sync::Semaphore for limits

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* Update client/rpc-spec-v2/src/chain_head/subscription/inner.rs

Co-authored-by: Sebastian Kunert <skunert49@gmail.com>

---------

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>
Co-authored-by: parity-processbot <>
Co-authored-by: Sebastian Kunert <skunert49@gmail.com>

snowmead committed:
Update Scheduler Pallet Documentation (#14740)

* Update pallet scheduler documentation, warning section, guidelines update

* Update call filter note

Co-authored-by: Kelvin Bonilla <bonilla_kelvin@hotmail.com>

* revert format cargo

* Doc wording

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* Doc wording

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

Ammend comments related to documentation

* Include additional warning section in `on_initialize` hook

* Amend doc

Co-authored-by: Sam Johnson <sam@durosoft.com>

* Amend doc

Co-authored-by: Sam Johnson <sam@durosoft.com>

* Move no_std to appropriate place

* Amend doc

Co-authored-by: Nate Armstrong <naterarmstrong@gmail.com>

* Amend comment

Co-authored-by: Nate Armstrong <naterarmstrong@gmail.com>

---------

Co-authored-by: Kelvin Bonilla <bonilla_kelvin@hotmail.com>
Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>
Co-authored-by: Sam Johnson <sam@durosoft.com>
Co-authored-by: Nate Armstrong <naterarmstrong@gmail.com>

acatangiu committed:
Revert "sc-cli: add no-beefy flag to cli config (#14754)" (#14766)

This reverts commit 2060c366fb1402deab188911551a43c3c41b36c0.

alvicsam committed:
Add `--workspace` to clippy (#14772)

* [DNM] Debug clippy

* add --workspace

altonen committed:
Make peer evictions less aggressive (#14619)

* Make peer evictions less aggressive

The original implementation of peer eviction prioritized aliveness over
connection stability which made the peer count unstable for some users.

As this may cause discomfort or infrastructure alerts if stability is
tracked, adjust the eviction to be less aggressive by only evicting
peers when the node has fully stalled. This causes the node to have some
peers who are inactive and won't send any block announcements.
These nodes are removed if the local node is able to receive at least
one block announcement from one of its peers as the inactivity of the
substream is detected when a notification is sent.

If the node won't send or receive any block annoucements for 30 seconds,
it's considered stalled and it will evict all peers,
causing `ProtocolController` to accept and establish connections from new
peers.

* Update client/network/sync/src/engine.rs

Co-authored-by: Dmitry Markin <dmitry@markin.tech>

* Track last send and received notification simultaneously

---------

Co-authored-by: Dmitry Markin <dmitry@markin.tech>
Co-authored-by: parity-processbot <>

nuke-web3 committed:
`cargo clippy +nightly --fix` run on downstream node template (#14693)

bkchr committed:
Improve `storage_alias` and make `UnlockAndUnreserveAllFunds` independent of the pallet (#14773)

* Make `storage_alias` more generic over the `prefix`

* Make `UnlockAndUnreserveAllFunds` indepenend from the pallet

* FMT

* Fix error reporting

* Rename prefix type

* Add test

* Apply suggestions from code review

Co-authored-by: Sam Johnson <sam@durosoft.com>

* ".git/.scripts/commands/fmt/fmt.sh"

---------

Co-authored-by: Sam Johnson <sam@durosoft.com>
Co-authored-by: command-bot <>

Commit:
[fix lint warnings: NFTs pallet] fix clippy::missing_docs_in_private_items warnings (#14610)

* add docs for impl_codec_bitflags

* add missing docs for type aliases

* add docs to transfer module

* add docs for settings module

* add docs to roles module

* add docs to metadata module

* add docs to migration module

* add missing docs to feature library

* methods not functions

* add docs to lock module

* add docs to attributes module

* add docs to create_delete_item module

* add docs for create_delete_collection module

* add docs to buy_sell module

* add missing doc for buy_sell module

* add docs to atomic_swap module

* add docs to atomic_swap module

* add docs for approvals module

* run cargo fmt

* Fix issues with multi-line comments

* Apply suggestions from code review

Co-authored-by: Jegor Sidorenko <5252494+jsidorenko@users.noreply.github.com>

* update from review

* fmt

* update from review

* remove bitflag example

* ".git/.scripts/commands/fmt/fmt.sh"

* Apply suggestions from code review

Co-authored-by: Squirrel <gilescope@gmail.com>

* add note about pallet features

---------

Co-authored-by: Jegor Sidorenko <jegor@parity.io>
Co-authored-by: Jegor Sidorenko <5252494+jsidorenko@users.noreply.github.com>
Co-authored-by: parity-processbot <>
Co-authored-by: Squirrel <gilescope@gmail.com>

brson committed:
Update wasm-opt to 0.114 (#14695)

altonen committed:
Revert "chore: update libp2p to 0.52.1 (#14429)" (#14722)

* Revert "chore: update libp2p to 0.52.1 (#14429)"

This reverts commit d38d176b844aab1338ce79eb71cd6df86c97d4a0.

* Fix dependencies

* Update dependencies

* Update Cargo.lock

liamaharon committed:
Free standing `elections-phragmen` and `tips` Gov V1 unlock/unreserve migrations (#14779)

* standalone elections-phragmen migration

* standalone tips migration

* remove redundant comment

liamaharon committed:
Disarm `OnRuntimeUpgrade::pre/post_upgrade` `Tuple` footgun (#14759)

* return error on incorrect tuple usage of pre_upgrade and post_upgrade

* add test

* comment lint

* Update frame/support/src/traits/hooks.rs

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* Update frame/support/src/traits/hooks.rs

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* address feedback

* Update frame/support/src/traits/hooks.rs

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* Update frame/support/src/traits/hooks.rs

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* Update frame/support/src/traits/hooks.rs

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* muharem comments

* Update frame/support/src/traits/hooks.rs

Co-authored-by: Muharem Ismailov <ismailov.m.h@gmail.com>

* Update frame/support/src/traits/hooks.rs

Co-authored-by: Muharem Ismailov <ismailov.m.h@gmail.com>

* remove useless type

---------

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>
Co-authored-by: Muharem Ismailov <ismailov.m.h@gmail.com>

liamaharon committed:
deprecate `try-runtime` subcommand and direct users to standalone cli (#14731)

* add deprecation notice

* remove try-runtime-cli tests

* update docs

* add estimated removal date

* deprecate and remove from node-cli and node-template

* try fix build script

* update comment

* fix link

* typo in build script

* Move `try-runtime-cli` install step outside of `check_dependent_project.sh` execution scope

* Update scripts/ci/gitlab/pipeline/build.yml

Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* remove chain arg

* build runtime with try-runtime feature

* kick ci

* kick ci

* use main branch

* specify sha in try-runtime-cli cargo install

* kick ci

* kick ci

---------

Co-authored-by: Vladimir Istyufeev <vladimir@parity.io>
Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

ashWhiteHat committed:
Runtime: avoid duplication and test all signature (#14663)

* runtime: all signature test

* test-utils: remove std duplication

* runtime: add bls verify test

bkchr committed:
Set `StateBackend::Transaction` to `PrefixedMemoryDB` (#14612)

* Yep

* Try to get it working everywhere

* Make `from_raw_storage` start with an empty db

* More fixes!

* Make everything compile

* Fix `child_storage_root`

* Fix after merge

* Cleanups

* Update primitives/state-machine/src/overlayed_changes/mod.rs

Co-authored-by: Davide Galassi <davxy@datawok.net>

* Review comments

* Fix issues

* Silence warning

* FMT

* Clippy

---------

Co-authored-by: Davide Galassi <davxy@datawok.net>

Ank4n committed:
remove page count requirement

Ank4n committed:
Merge branch 'master' into ankan/paged-rewards-rebased2

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3416775

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3416774

Ank4n committed:
paging is always enabled

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3416880

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3416879

Ank4n commented:
bot fmt

Commit:
".git/.scripts/commands/fmt/fmt.sh"

Ank4n committed:
no clipping or sorting of nominators

Ank4n commented on bin/node/runtime/src/lib.rs:
```diff
@@ -416,7 +416,7 @@ impl pallet_babe::Config for Runtime {
 	type WeightInfo = ();
 	type MaxAuthorities = MaxAuthorities;
-	type MaxNominators = MaxNominatorRewardedPerValidator;
+	type MaxNominators = MaxExposurePageSize;
```
This is used for weight calculation for reporting equivocation. The older configuration (changed [here](https://github.com/paritytech/substrate/pull/14471)) was incorrect since for slashing, the max nominators exposed could be larger than `MaxNominatorRewardedPerValidator `. `MaxNominatorRewardedPerValidator` is only used to limit nominators that can be rewarded but nominators eligible for slashing (per validator) is unbounded (though there is a hard count for all nominators => 22500).  

Not sure the correct way to solve this and is most probably out of scope of this PR. Setting this to `MaxExposurePageSize` is strictly same as before (though now its more apparent that this configuration is wrong).

Ank4n committed:
small edits to the changelog
